### PR TITLE
8231622: SuppressWarning("serial") ignored on field serialVersionUID

### PR DESCRIPTION
**Pull request by lgxbslgx**
Hi all,

Currently, a warning is generated when compiling the following code by using the option `-Xlint`.

```
import java.io.Serializable;
class B implements Serializable {
    @SuppressWarnings("serial")
    private static final byte serialVersionUID = 5;
}

output:
B.java:4: warning: [serial] serialVersionUID must be of type long in class B
    private static final byte serialVersionUID = 5;
                              ^
1 warning
```

The annotation `@SuppressWarnings("serial")` on  field serialVersionUID is ignored by the compiler.
This patch fixes it and adds some tests.
Thank you for taking the time to review.

Best Regards.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8231622](https://bugs.openjdk.java.net/browse/JDK-8231622): SuppressWarning("serial") ignored on field serialVersionUID

### Reviewers
 * [Jonathan Gibbons](https://openjdk.java.net/census#jjg) (@jonathan-gibbons - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1626/head:pull/1626`
`$ git checkout pull/1626`

### CODE REVIEW

**Review by jonathan-gibbons — CHANGES_REQUESTED** [review]
For tests that use `/ref=file` `-XDrawDiagnostics` the conventions are:

1. Omit the complete legal header, including the copyright and license
2. After `@test` add the text `/nodynamiccopyright/`

#1 protects the file against any future changes in the length of the legal header, that might affect line numbers, and #2 is for use by automated scripts that may check for the presence of the legal header.

**Comment by lgxbslgx** [comment]
Thank you for your comment. I revise my code according to your suggestion.

And I found some previous tests use the wrong style which are like this patch. I want to modify them so that these tests won't mislead other developers, especially the new developers. Could I get your help to open a new issue about it in the bug tracker? Thanks a lot.

**Comment by jonathan-gibbons** [comment]
> Thank you for your comment. I revise my code according to your suggestion.
> 
> And I found some previous tests use the wrong style which are like this patch. I want to modify them so that these tests won't mislead other developers, especially the new developers. Could I get your help to open a new issue about it in the bug tracker? Thanks a lot.

Can you give me a list of tests that you think need to be addressed?

**Review by jonathan-gibbons — APPROVED** [review]
I approve these changes.

That being said, a more modern idiom for negative tests involving small source files and corresponding small golden files, is to write a single file that uses library code like `toolbox.ToolBox` to write files on the fly, perhaps from content in a text block, and to compare the output against another text block.  For slightly more advanced usage, you can compile strings directly using API like `SimpleJavaFileObject`, and avoid the file system altogether.

**Comment by lgxbslgx** [comment]
> Can you give me a list of tests that you think need to be addressed?

I write a python script to help me to find the possible wrong style tests. The script finds string `Copyright`, `ref=` and `-XDrawDiagnostics` in the files of directory `test`. If a file contains these three strings, it will print the full path name of the file to `result.txt`.

As you can see the entries in the annex `result.txt`, for example, the `test/langtools/tools/doclint/multipackage/MultiPackage.java` is not a wrong style test and `test/langtools/tools/javac/AnonStaticMember_1.java` is a wrong style test.
Later, I will check all the entries manually and revise the wrong style tests because I can't find a better way to do that.

The content of the python script is shown below. I don't attach it as an annex because GitHub doesn't support the python file type.
```
#!/usr/bin/python
import os


def clean(output_file):
    if os.path.exists(output_file):
        os.remove(output_file)


def outputFileName(full_path, output_file):
    output_file = open(output_file, 'a')
    output_file.write(full_path)
    output_file.write("\n")
    output_file.close()


def findStrListInTestFile(full_path, find_str_list):
    search_file = open(full_path, 'r')
    content = search_file.read()
    search_file.close()
    for find_str in find_str_list:
        is_found = content.find(find_str)
        if is_found == -1:
            return False
    return True


def findWrongStyleTests(find_dir, find_str_list, output_file):
    for dir_path, dir_names, file_names in os.walk(find_dir):
        for file_name in file_names:
            full_path = os.path.join(dir_path, file_name)
            is_found = findStrListInTestFile(full_path, find_str_list)
            if is_found:
                outputFileName(full_path, output_file)


if __name__ == "__main__":
    findDir = "test"
    findStrList = ["Copyright", "ref=", "-XDrawDiagnostics"]
    outputFile = "result.txt"
    clean(outputFile)
    findWrongStyleTests(findDir, findStrList, outputFile)

```

The steps to run the script are shown below.
- Enter the jdk main directory: `cd path/to/jdk`
- Create a new python script: `vim test.py`
- Copy the code above to the file `test.py`
- Use python command to run the script: `python test.py`

The file `result.txt` is shown below.
[result.txt](https://github.com/openjdk/jdk/files/5656821/result.txt)

Anyway, I don't think we should continue to discuss the concrete content in this patch.

**Comment by lgxbslgx** [comment]
@jonathan-gibbons I revise the test code by using the library toolbox. Thank you for taking the time to review.

File: test/langtools/tools/javac/T8231622/T8231622.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+                .writeAll()
+                .getOutputLines(Task.OutputKind.DIRECT);
+
+        List<String> expected = Arrays.asList(
+                "T8231622_1.java:3:29: compiler.warn.long.SVUID: T8231622_1",
+                "1 warning");
+
+        if (!Objects.equals(output, expected)) {
+            throw new AssertionError("incorrect output\nactual=" + output + "\nexpected=" + expected);
+        }
```
Comment (jonathan-gibbons) [inline test/langtools/tools/javac/T8231622/T8231622.java:79]: ToolBox provides `checkEquals` which provides a more detailed error message in the case of a difference.

**Review by jonathan-gibbons — COMMENTED** [review]
Nice improvement. You can make it even better in 3 places by using `ToolBox.checkEquals`, which provides more detailed reporting in the case of a difference being found.

File: test/langtools/tools/javac/T8231622/T8231622.java (hunk 1 above)
Comment (lgxbslgx) [reply test/langtools/tools/javac/T8231622/T8231622.java]: Thank you for the suggestion. I fixed it.

**Comment by lgxbslgx** [comment]
> Anyway, I don't think we should continue to discuss the concrete content in this patch.

@jonathan-gibbons I submit a patch[1] to solve the previous bad style tests. We can continue the discussion there.

[1] https://github.com/openjdk/jdk/pull/1732

**Review by jonathan-gibbons — APPROVED** [review]
Latest change looks good

**Comment by lgxbslgx** [comment]
Could I ask your help to sponsor this patch? Thanks a lot.

**Pull request closed without merging**